cmd/kubectl-gadget/utils.go

```diff
@@ -131,3 +155,225 @@ func randomTraceID() string {
+	//
+	// watch, err := traceRestClient.
+	//	Get().
+	//	Namespace("gadget").
+	//	Resource("traces").
+	//	VersionedParams(&listTracesOptions, scheme.ParameterCodec).
+	//	Watch(context.TODO())
+	// if err != nil {
+	//	contextLogger.Fatalf("Error waiting for traces: %q", err)
+	// }
+	// for event := range watch.ResultChan() {
+	//	fmt.Printf("Event: %v\n", event)
+	//	if data, ok := event.Object.(*metav1.Status); ok {
+	//		contextLogger.Infof("watcher status: %s", data.Message)
+	//	} else if t, ok := event.Object.(*gadgetv1alpha1.Trace); ok {
+	//		fmt.Printf("Got: %v\n", t)
+	//	} else {
+	//		contextLogger.Fatalf("Error waiting for traces: got unexpected %v", event)
+	//	}
+	// }
+
+	var results gadgetv1alpha1.TraceList
+	start := time.Now()
+RetryLoop:
+	for {
+		results = gadgetv1alpha1.TraceList{}
+		err = traceRestClient.
+			Get().
+			Namespace("gadget").
+			Resource("traces").
+			VersionedParams(&listTracesOptions, scheme.ParameterCodec).
+			Do(context.TODO()).
+			Into(&results)
+		if err != nil {
+			deleteTraces(contextLogger, traceRestClient, traceID)
+			contextLogger.Fatalf("Error getting traces: %q", err)
+		}
+
+		timeout := time.Now().Sub(start) > 2*time.Second
+		successNodeCount := 0
+		nodeErrors := make(map[string]string)
+		for _, i := range results.Items {
+			if i.Status.State == "Completed" || i.Status.State == "Started" {
+				successNodeCount++
+			} else {
+				if timeout {
+					nodeErrors[i.Spec.Node] = i.Status.OperationError
+					continue
+				}
+				time.Sleep(100 * time.Millisecond)
+				continue RetryLoop
+			}
+		}
+		for node, err := range nodeErrors {
+			contextLogger.Warningf("Error getting traces from node %q: %s", node, err)
+		}
+		if successNodeCount == 0 {
+			deleteTraces(contextLogger, traceRestClient, traceID)
+			contextLogger.Fatalf("Error getting traces from all nodes")
+		}
+		break RetryLoop
+	}
+
+	if callback == nil {
+		genericStreamsDisplay(contextLogger, client, params, &results)
+	} else {
+		callback(contextLogger, nodes, &results)
+	}
+	deleteTraces(contextLogger, traceRestClient, traceID)
+}
+
+func genericStreamsDisplay(contextLogger *log.Entry, client *kubernetes.Clientset, params *utils.CommonFlags, results *gadgetv1alpha1.TraceList) {
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	failure := make(chan string)
+
+	for _, i := range results.Items {
+		if params.Node != "" && i.Spec.Node != params.Node {
+			continue
+		}
+		go func(nodeName, namespace, name string) {
+			cmd := fmt.Sprintf("exec gadgettracermanager -call receive-stream -tracerid trace_%s_%s",
+				namespace, name)
+			var err error
+			err = execPod(client, nodeName, cmd, os.Stdout, os.Stderr)
+			if fmt.Sprintf("%s", err) != "command terminated with exit code 137" {
+				failure <- fmt.Sprintf("Error running command: %v\n", err)
+			}
+		}(i.Spec.Node, i.ObjectMeta.Namespace, i.ObjectMeta.Name) // i is invalidated by the above for loop, causes races
```

Comment:
I never understood that comment. Are those races a problem? should we create a copy of `i` before calling the func to avoid those races?

Reply:
The comment is confusing, I'll remove it. As it is written, there are no race conditions. It was just an explanation why we pass the variables go routine as parameters to the go routine instead of accessing `i` directly.

Direct access in the Go routine would be racy but I see that this version of Go does not allow this code anyway.
https://play.golang.org/p/Jr6P6XlRg50

Reply:
But the Go runtime does not always detect the problem:
https://play.golang.org/p/8VVxCUYPm6h